src/odemis/gui/comp/overlay/shapes.py

```diff
@@ -180,39 +180,37 @@ def on_leave(self, evt):
         else:
             WorldOverlay.on_leave(self, evt)
 
-    def _activate_shapes(self, flag=True):
-        """Activate or de-activate the shapes."""
+    def _deselect_shapes(self):
+        """Deselect shapes except the selected shape."""
         for shape in self._shapes:
-            shape.active.value = flag
+            if shape.selected.value and shape != self._selected_shape:
+                shape.selected.value = False
 
     def _on_tool(self, selected_tool):
         """Update the overlay when it's active and tools change."""
-        if selected_tool == self.tool:
-            self.active.value = True
-            self._activate_shapes(True)
-        else:
-            self.active.value = False
-            self._activate_shapes(False)
-            self.cnvs.reset_default_cursor()
+        self.active.value = selected_tool == self.tool
 
-    def _get_shape(self, evt):
+    def _get_shape(self, v_pos: Tuple[float, float]):
         """
-        Find a shape corresponding to the given on_left_down event position.
+        Find a shape corresponding to the given view position.
         :return: the most appropriate shape.
             If no shape is found, it returns None.
         """
         if self._shapes:
-            pos = self.cnvs.view_to_phys(evt.Position, self.cnvs.get_half_buffer_size())
-            for shape in self._shapes[::-1]:
-                if shape.is_point_in_shape(pos):
-                    return shape
+            v_centers = numpy.array([shape.v_center for shape in self._shapes])
+            v_centers_kdtree = cKDTree(v_centers)
+            distances, indices = v_centers_kdtree.query(numpy.array([v_pos[0], v_pos[1]]), k=4)
+            for distance, index in zip(distances, indices):
```

Comment:
If I understand correctly, it'll only look at the 4 shapes with the centers the closest to the v_pos. If none of them is over the point, then it's considered that no shape is over the point? Is that correct? 
If so, that's just an heuristic, relying on the fact that most shapes are about the same area. If there is one very large shape, this heuristic could easily fail, right?
If so, I'd suggest to at least document this limitation. Ideally, you'd ensure that you *always* find a shape if there is  one that covers the point.

Also, wouldn't it be simpler, instead of using KDTree on the centers, to keep track of the bounding box of each shape, and use that BBox to quickly discard far away shapes?

Reply:
I will mention the limitations